### PR TITLE
[FSM] Remove SingleBlockImplicitTerminator from StateOp, TransitionOp

### PR DESCRIPTION
**Pull request by mortbopet**
This was enforcing unintended ops which shows up when generating FSM IR from a frontend. Previously, `SingleBlockImplicitTerminator<"OutputOp">` required an OutputOp to be in both the output and transition regions, where it really only made sense for the output region.

It's a similar story for the TransitionOp and `SingleBlockImplicitTerminator<"ReturnOp">`.

This, and other small fixes, have been moved to FSMOps.cpp

### CODE REVIEW

**Comment by mikeurbach** [comment]
I don't know where this op's design is coming from or where it is going, but I don't think these terminators are "unintended".

The description of StateOp and TransitionOp both explicitly mention these terminators and their uses.

Is your intent for StateOp to no longer use OutputOp to indicate the machine's outputs in the given state? And similarly, not use the TransitionOp's terminating ReturnOp to indicate that transition's guard condition?

At the very least, I would hope to see the descriptions updated to make the new semantics explicit,

**Comment by mortbopet** [comment]
> Is your intent for StateOp to no longer use OutputOp to indicate the machine's outputs in the given state? And similarly, not use the TransitionOp's terminating ReturnOp to indicate that transition's guard condition?

The intent is to keep these semantics exactly as is. However, that is now what was implemented with `SingleBlockImplicitTerminator`.  
The `SingleBlockImplicitTerminator` constraint was being enforced for both regions of either op, where in reality, the constraint only applies to one of them - `StateOp` needs its `output` region to be terminated by an `OutputOp` and `TransitionOp` only needs its `guard` region to be terminated by a `ReturnOp`.

This results in IR that looks like this "under the hood" (implicit terminators shows):
```mlir
fsm.machine @foo(%arg0: i1) -> i1 attributes {initialState = "IDLE"} {
  fsm.state "IDLE" output  {
    %true = arith.constant true
    fsm.output %true : i1
  } transitions  {
    fsm.transition @BUSY guard  {
      fsm.return %arg0
    } action  {
      ...
      fsm.return // <- implicit, unintended
    }
    ...
    fsm.output // <- implicit, unintended
  }
}
```

Previously, these ops were generated by the `ensureTerminator` calls in the `build` functions of the ops, but a front-end like PyCDE has to manually create these out-of-place terminators.

> At the very least, I would hope to see the descriptions updated to make the new semantics explicit

In short, no semantics changed. In fact, the descriptions are now more accurate of the respective ops since they didn't reference the (unintended) terminator constraint on the 2nd region which each op defines.

**Comment by mikeurbach** [comment]
Ok thanks for the explanation. I guess I missed that these ops had multiple regions and the SingleBlockImplicitTerminator applied to both.

**Comment by mortbopet** [comment]
I'll go ahead and merge this as it's blocking further FSM work wrt. PyCDE and HW lowering, and by the fact that nothing else in CIRCT (atm) use this.
